xwiki-platform-core/xwiki-platform-yjs/xwiki-platform-yjs-websocket/src/main/java/org/xwiki/yjs/websocket/internal/Room.java

```diff
@@ -0,0 +1,171 @@
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.yjs.websocket.internal;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import jakarta.inject.Inject;
+import jakarta.websocket.Session;
+
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.annotation.InstantiationStrategy;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.CodedInputStream;
+import com.google.protobuf.CodedOutputStream;
+
+import static org.apache.commons.lang3.exception.ExceptionUtils.getRootCauseMessage;
+import static org.xwiki.component.descriptor.ComponentInstantiationStrategy.PER_LOOKUP;
+
+/**
+ * This object stored the list of currently active sessions for a room. A room is usually containing the users editing a
+ * given document in realtime.
+ *
+ * @version $Id$
+ * @since 17.6.0RC1
+ */
+@Component(roles = Room.class)
+@InstantiationStrategy(PER_LOOKUP)
+public class Room
+{
+    @Inject
+    private Logger logger;
+
+    private Runnable clearRoom;
+
+    private final Map<Session, Long> sessionIds = new ConcurrentHashMap<>();
+
+    /**
+     * Initialize the room. This method must be call directly after instantiation.
+     *
+     * @param clearRoom a clear room runnable action, that needs to be called once the room becomes empty again.
+     */
+    public void init(Runnable clearRoom)
+    {
+        this.clearRoom = clearRoom;
+    }
+
+    /**
+     * Disconnect a session from the room.
+     *
+     * @param session the session to disconnect.
+     */
+    public void disconnect(Session session)
+    {
+        Long clientID = this.sessionIds.get(session);
+        if (clientID == null) {
+            return;
+        }
+        try {
+            this.broadcast(session, buildDisconnectMessage(clientID));
+        } catch (IOException e) {
+            this.logger.warn("Failed to build the disconnect message for client id [{}]. Cause: [{}]", clientID,
```

Comment:
The failure could be also when broadcasting the disconnect message no? I'd make the warning message a bit more generic, e.g. "Failed to broadcast the disconnect message from client [{}]".